content/blog/2022/05/verifying-image-signatures.md

```diff
@@ -0,0 +1,128 @@
+new [verify-image-signatures](https://github.com/kubewarden/verify-image-signatures) policy
+
+If you don't know how Sigstore works, we recommend reading our
+previous [post](https://www.kubewarden.io/blog/2022/04/securing-kubewarden-policies/)
+
+## Verify Image Signatures Policy
+
+This policy validates Pods by checking their container images for signatures (that is, containers, init containers and 
+ephemeral containers in the pod). If all signature validations pass or there is no container image that matches the image
+name query, the Pod will be accepted.
+
+Once the image has been verified with a signature, one needs to ensure that the specific image tag from that signature 
+is the one instantiated. This is achieved by the policy by mutating the Pod request and adding that validated image digest. 
+This mutation can be disabled in the [settings](https://github.com/kubewarden/verify-image-signatures#settings).
+
+## Let's see it in action!
+
+For this example, a Kubernetes cluster with Kubewarden already installed is required. The installation process is
+described in the [quick start guide](https://docs.kubewarden.io/quick-start.html).
+
+We need an image with a signature that we can verify. You can use [cosign](https://github.com/sigstore/cosign) to sign
+your images. For this example we'll use the image
+`ghcr.io/viccuad/app-example:v0.1.0` that was signed using keyless verification. 
+
+Obtain the issuer and subject using cosign.
+
+```
+COSIGN_EXPERIMENTAL=1 cosign verify ghcr.io/viccuad/app-example:v0.1.0
+```
+
+```
+...
+"Issuer": "https://github.com/login/oauth",
+"Subject": "me@viccuad.me"
+...
+```
+
+Let's create a cluster-wide policy that will verify all images, and
+we'll use the issuer and subject for verification:
+
+```
+kubectl apply -f - <<EOF
+apiVersion: policies.kubewarden.io/v1alpha2
+kind: ClusterAdmissionPolicy
+metadata:
+  name: verify-image-signatures
+spec:
+  module: ghcr.io/kubewarden/policies/verify-image-signatures:v0.1.4
+  rules:
+  - apiGroups: [""]
+    apiVersions: ["v1"]
+    resources: ["pods"]
+    operations:
+    - CREATE
+    - UPDATE
+  mutating: true
+  settings:
+    signatures:
+      - image: "*"
+        keyless:
+          - issuer: "https://github.com/login/oauth"
+            subject: "me@viccuad.me"
+EOF
+````
+
+Wait for the policy to be active:
+
+```
+kubectl wait --for=condition=PolicyActive clusteradmissionpolicies verify-image-signatures
+```
+
+Verify we can create pods with containers that are signed:
+
+```
+kubectl apply -f - <<EOF
+apiVersion: v1
+kind: Pod
+metadata:
+  name: verify-image-valid
+spec:
+  containers:
+  - name: test-verify-image
+    image: ghcr.io/viccuad/app-example:v0.1.0
+EOF
+```
+
+Then check that the image was modified with the digest
+
+```
+kubectl get pod verify-image-valid -o=jsonpath='{.spec.containers[0].image}'
+```
+
+```
+ghcr.io/viccuad/app-example:v0.1.0@sha256:d97d00f668dc5b7f0af65edbff6b37924c8e9b1edfc0ab0f7d2e522cab162d38
```

Comment:
This is not a valid reference, it should be:

```
ghcr.io/viccuad/app-example@sha256:d97d00f668dc5b7f0af65edbff6b37924c8e9b1edfc0ab0f7d2e522cab162d38
```

Is that a typo or is the policy generating a broken ref?

Reply:
It's ok to leave the tag, that's a valid reference. You can check with 
```
docker pull ghcr.io/viccuad/app-example:v0.1.0@sha256:d97d00f668dc5b7f0af65edbff6b37924c8e9b1edfc0ab0f7d2e522cab162d38
```
I think it is better to leave the tag, so users know which version is deployed. What do you think?

Reply:
Wow, I didn't know that was a valid image reference! Leave it there then!